reVX/offshore/dist_to_ports.py

```diff
@@ -0,0 +1,395 @@
+        else:
+            logger.info('Computing least cost distance to ports in serial')
+            for i, port in self.ports.iterrows():
+                port_idx = port[['row', 'col']].values
+                port_dist = port['dist_to_pixel']
+                dist = self._lc_dist_to_port(self.cost_arr, port_idx,
+                                             port_dist)
+                dist_to_ports = np.minimum(dist_to_ports, dist)
+                logger.debug('Computed least cost distance for {} of {} '
+                             'ports'.format((i + 1), n_ports))
+
+        # Set onshore pixels least cost distance to -1
+        dist_to_ports[self.cost_arr > 90] = -1
+
+        return dist_to_ports
+
+    def save_as_layer(self, layer_name, dist_to_ports, chunks=(128, 128)):
+        """
+        Save distance to ports as an exclusion layer within excl_fpath
+
+        Parameters
+        ----------
+        layer_name : str
+            Name of layer under which distance to ports is to be saved
+        dist_to_ports : ndarray
+            Distance to ports data
+        chunks : tuple
+            Chunk size of dataset in .h5 file
+        """
+        if len(dist_to_ports.shape) < 3:
+            dist_to_ports = np.expand_dims(dist_to_ports, 0)
+
+        logger.info('Saving {} to {}'.format(layer_name, self._excl_fpath))
+        description = ("Minimum distance to the nearest {} in meters, onshore "
+                       "pixels have a value of -1".format(layer_name))
+        ExclusionsConverter._write_layer(self._excl_fpath, layer_name,
+                                         self._profile, dist_to_ports,
+                                         chunks=chunks,
+                                         description=description)
+
+    @classmethod
+    def run(cls, ports, excl_fpath, dist_layer='dist_to_coast',
+            ports_layer=None, chunks=(128, 128), max_workers=None,
+            update_layer=True):
+        """
+        Compute the least cost distance to the nearest ports in km, either
+        write as a new layer in the exclusion .h5 file or update an existing
+        layer
+
+        Parameters
+        ----------
+        ports : str
+            Path to shape file containing ports to compute least cost distance
+            to
+        excl_fpath: str
+            Path to exclusions .h5 file with distance to coast layer. Will also
+            be the file into which the least cost distance to port is saved.
+        dist_layer : str, optional
+            Exclusions layer with distance to coast values,
+            by default 'dist_to_coast'
+        ports_layer : str, optional
+            Exclusion layer under which the distance to ports layer should be
+            saved, if None use the ports file-name, by default None
+        chunks : tuple, optional
+            Chunk size of dataset in .h5 file, by default (128, 128)
+        max_workers : int, optional
```

Comment:
The parallelization scheme for this is by port right? So if adding one new port to the layer, max_workers should be 1?

Reply:
yes, but if your run with update_layer it could have multiple new ports in the shape file (or future .csv)